Native `rearrange` in `functorch`

ringohoffman opened the pull request:
Fixes #92675

Here we implement a native version of [`einops.rearrange`](https://einops.rocks/api/rearrange/) using first class dims to perform the operations. The string parsing + validation, documentation, and relevant tests are adapted from `einops`. The API is exactly the same as the `einops` API.

The main idea is to take the string and convert it to a left and right `ParsedExpression`, and then find a mapping from the axes to first class dims. Once the mapping exists we convert the left expression `composition` list into a `Tensor.__getitem__` index and the right expression `composition` into the `Tensor.order` arguments, and then use this to dynamically create a callable that performs the `rearrange` operation as specified by the pattern.

cc: @ezyang

Commit:
Adapt einops.rearrange to use torch first class dims

The parsing code is taken from einops; this code doesn't appear to be fully functional with AnonymousAxis quite yet

Commit:
Add adapted einops tests for rearrange and parsing

These are taken from einops, but adapted to use unittest and torch

Commit:
Add failing test for squeeze

Commit:
Initial fix for squeeze

It's convenient for us to allow for a AnonymousAxis of size 1 since  we need to map them to dims still

ringohoffman commented:
@ezyang I have 1 failing test that I'm not sure what to make of. When I try to perform a squeeze, I get a `functorch.dim.Tensor` back instead of a `torch.Tensor`.

```python
import torch
from functorch._C import dim as _C
dims = _C.dims

x = x = torch.randn((2, 1, 3, 4, 1, 5))
c1, c2, c3, c4, c5, c6 = dims(len(x.shape))
ans = x[c1, c2, c3, c4, c5, c6].order(c1, c3, c4, c6)
type(ans)  # functorch.dim.Tensor
```
Is this expected behavior? How is it recommended to perform a `squeeze` with first class dims?

Commit:
Use permalink in URL

Skylion007 commented on functorch/dim/einops/rearrange.py:
```diff
@@ -0,0 +1,150 @@
+    try:
+        left_str, right_str = pattern.split("->")
+    except ValueError:
+        raise ValueError("Pattern must contain a single '->' separator")
```
Either raise from the original ValueError or raise from None

ringohoffman replied:
I think `not enough values to unpack`/`too many values to unpack` is not very helpful here, so I think I will `raise from None` thanks for the suggestion.

Commit:
Raise from None to suppress unhelpful parsing error

not enough values to unpack / too many values to unpack is not a very helpful error message to users of rearrange, so I think we should suppress this error message

Commit:
Add functorch.dim.Tensor.sum over anonymous axes to convert result to torch.Tensor

Commit:
Add codeowner header

ringohoffman commented:
> @ezyang I have 1 failing test that I'm not sure what to make of. When I try to perform a squeeze, I get a `functorch.dim.Tensor` back instead of a `torch.Tensor`.
> 
> ```python
> import torch
> from functorch._C import dim as _C
> dims = _C.dims
> 
> x = x = torch.randn((2, 1, 3, 4, 1, 5))
> c1, c2, c3, c4, c5, c6 = dims(len(x.shape))
> ans = x[c1, c2, c3, c4, c5, c6].order(c1, c3, c4, c6)
> type(ans)  # functorch.dim.Tensor
> ```
> 
> Is this expected behavior? How is it recommended to perform a `squeeze` with first class dims?

@ezyang I fixed it by summing along the anonymous axes 856ad004fe64a3a45dbe904cf264a8fbac29f0a9. Not sure if there is a better way, but this PR is ready for review.

ringohoffman commented:
I'm working on memoizing the string parsing

Commit:
Use double quotes for consistent style

improve, add docstrings

Commit:
Memoize pattern parsing

Since the an equivalent result is computed for tensors with the same number of dimensions, with the same pattern and specified axes lengths, the pattern parsing can be memoized.

ezyang commented:
@zdevito is going to take a look at this, he is more familiar with torchdim than I

zdevito commented:
Rebinding dims has a tricky implementation because dimensions can be long-lived (e.g. a whole self-attention), and it would break some invariants if the size of dimension could change while tensors referring to it still exist.

One approach to fast memoization that would be relatively simple would be to generate and eval a small function that corresponds to each string:

rearrange("a b c d e -> (a b) c d e") becomes: 

```
def generated(i0):
   a, b, c, d, e = dims(5)
   return i0[a, b, c, d, e].order((a, b), c, d, e)
```

This avoids having to do re-indexing, and generating the dims themselves is pretty fast (dims is one C API call).

Commit:
Dynamically generate and cache callables that perform rearrange based on a tensor's number of dims and patterns

@zdevito's suggestion: https://github.com/pytorch/pytorch/pull/101957#issuecomment-1561374663

ringohoffman commented:
@zdevito LMK what you think of this implementation; it's finally faster than `einops` (22% on the aforementioned example)!

@ezyang expressed concerns with using `exec` on the previous implementation: https://github.com/pytorch/pytorch/pull/93222#discussion_r1092569078

Commit:
Move rearrange callable creation logic to rearrange

refresh stale docstrings

Commit:
Validate axes_lengths values

We assert that the type of the values of axes_lengths must be exactly int, since we could theoretically execute arbitrary code if, for example, we pass in a class whose __repr__ is Python code that calls exec

Commit:
Make axes_lengths required for validation functions

If we make it non-variadic, we can be more sure that they will not accidentally not be passed in when it is very important that they are

Commit:
Clean up docs, improve formatting

zdevito commented:
I'm glad that code generation is able to make it fast enough!

I wonder if some of the handling of `...` can be simplified from `dimlists` which are the analogous concept in first-class dimensions. They aren't a document feature yet, but maybe they can help here. An example:

```
from functorch.dim import dims, dimlists
import torch

A = torch.rand(3, 4, 5)

x = dims(1)
y_ = dimlists(1)
# ..., x -> x, ...
print(A[y_, x].order(x, y_).size())
```

Commit:
Handle case of identity operation on 0-dimension tensor

Commit:
Add validations for dimension mismatch

These would have errored out anyway, but we can provide a more helpful, specific error message here by comparing the expected number of dims to what we received

Commit:
Clean up code generation with helper function

ringohoffman commented:
> I'm glad that code generation is able to make it fast enough!
> 
> I wonder if some of the handling of `...` can be simplified from `dimlists` which are the analogous concept in first-class dimensions. They aren't a document feature yet, but maybe they can help here. An example:
> 
> ```
> from functorch.dim import dims, dimlists
> import torch
> 
> A = torch.rand(3, 4, 5)
> 
> x = dims(1)
> y_ = dimlists(1)
> # ..., x -> x, ...
> print(A[y_, x].order(x, y_).size())
> ```

it's great to know about `dimlists`, but

1. (IMO) implementing it makes the handling more complicated:  
2. I'm running into some unexpected issues with it that I wasn't when manually handling ellipsis...

```python
import torch
from functorch._C import dim as _C
from functorch.dim import rearrange

dims, dimlists = _C.dims, _C.dimlists

x = torch.zeros([1, 1, 1, 1, 1])

rearrange(x, "... -> (...)")
# generates:
def do_rearrange1(tensor):
    dx = dimlists(1)
    tensor = tensor[dx].order((dx,))  # DimensionBindError: DimList not bound
    return tensor


rearrange(x, "... -> ...")
# generates:
def do_rearrange2(tensor):
    dx = dimlists(1)
    tensor = tensor[dx].order(dx)  # DimensionBindError: DimList not bound
    return tensor

```

zdevito commented:
If it doesn't help, it's not necessary to use dimlist, I was just thinking it would prevent having to recompile the function for different tensor dimensionalities. 

There is a bug in the dimlist handling causing the the DimList not bound error due to a corner case. It only applies if dx is the only argument . `tensor[(dx,)]` is a workaround. Another limitation is that if you want to flatten the dimlist into single dimension it needs to look like `(*dx,)` instead of `(dx,)`. So the first example would be:

```
tensor = tensor[(dx,)].order((*dx,)) 
```

This corner case is one of the reasons they aren't documented yet.

Commit:
Add unit tests for parsing validation

ringohoffman commented:
@zdevito please LMK what else you need to see for this to be merge-able

Commit:
Merge branch 'main' into functorch-rearrange

zdevito commented:
Just got back from vacation over Memorial Day. I should be able to take a detailed look tomorrow and get back to you.

zdevito commented on functorch/dim/einops/_parsing.py:
```diff
@@ -0,0 +1,217 @@
+            raise TypeError("Axis lengths must be integers")
+
+    if left.has_non_unitary_anonymous_axes or right.has_non_unitary_anonymous_axes:
+        raise ValueError("Non-unitary anonymous axes are not supported in rearrange (exception is length 1)")
```
Maybe clarify: "rearrange only supports unnamed axes of size 1"

zdevito commented on functorch/dim/einops/_parsing.py:
```diff
@@ -0,0 +1,217 @@
+        raise ValueError(f"Identifiers not found in expression: {unmatched_axes}")
+
+
+def comma_sepparate(iterable: Iterable[Any]) -> str:
```
separate

zdevito commented on functorch/dim/einops/rearrange.py:
```diff
@@ -0,0 +1,184 @@
+            f"def {custom_rearrange_callable_name}(tensor):\n"
+            f"    {comma_sepparate(first_class_dims)} = dims({n_dims})\n"
+            + (
+                f"    for dim, length in {specified_lengths}:\n"
```
Might be slightly fewer byte codes if generated unrolled as:
```
a.size = 3
b.size = 4
```

zdevito commented on functorch/dim/einops/rearrange.py:
```diff
@@ -0,0 +1,184 @@
+        # an identity rearrangement on a 0-dimension tensor
+        return lambda tensor: tensor
+
+    first_class_dims: Tuple[Dim, ...] = (dims(n_dims),) if n_dims == 1 else dims(n_dims)
```
What is the reasoning to generate Dim objects during code generation? It might be more robust to use strings to refer to the variables that will become Dim objects so that this code doesn't rely on the code for turning a Dim in to a string, which is designed only for debugging.

ringohoffman replied:
It was a remnant of an earlier iteration, but I kept it in because it was convenient to rely on the `Dim.__str__`. I can modify this to not rely on it.

zdevito reviewed (commented):
Overall looks good! I have a few inline comments, but the overall approach should work.

ringohoffman commented on functorch/dim/__init__.py:
```diff
@@ -2,6 +2,7 @@
 from typing import Union, Sequence
 import inspect
 import dis
+from .einops import rearrange as rearrange
```
Do you want `einops` or `rearrange` imported here so that ppl can do

```python
import functorch

functorch.rearrange
# or
functorch.einops.rearrange
```

?

ringohoffman commented on test/functorch/test_rearrange.py:
```diff
@@ -0,0 +1,160 @@
+        x = torch.zeros([1, 1, 1, 1, 1])
+        rearrange(x, 'a b c d ... ->  a b c ... d')
+        with self.assertRaises(ValueError):
+            rearrange(x, 'a b c d (...) ->  a b c ... d')
```
That is correct.

ringohoffman replied:
https://github.com/arogozhnikov/einops/blob/230ac1526c1f42c9e1f7373912c7f8047496df11/einops/einops.py#LL259C1-L260C1

https://github.com/pytorch/pytorch/blob/3fb4003bba69b5b07105661be9fc2d27cc477877/functorch/dim/einops/_parsing.py#L185

Commit:
Move functorch.dim.einops to functorch.einops

https://github.com/pytorch/pytorch/pull/101957#discussion_r1213428721

Commit:
Fix typo in separate

https://github.com/pytorch/pytorch/pull/101957#discussion_r1213455103

Commit:
Clarify rearrange expression validation error messages

https://github.com/pytorch/pytorch/pull/101957#discussion_r1213444893

Commit:
Add MIT license from einops

https://github.com/pytorch/pytorch/pull/101957#discussion_r1213437604

ringohoffman commented on functorch/dim/einops/_parsing.py:
```diff
@@ -0,0 +1,217 @@
+"""Adapted from https://github.com/arogozhnikov/einops/blob/36c7bb16e57d6e57f8f3050f9e07abdf3f00469f/einops/parsing.py."""
```
LMK if what I did is what you had in mind

Commit:
Unroll axis length assignment loop

https://github.com/pytorch/pytorch/pull/101957#discussion_r1213462134

Commit:
Refactor rearrange to use str representations of first class dims instead of converting first class dims to str

https://github.com/pytorch/pytorch/pull/101957#discussion_r1213463598

Commit:
Merge branch 'main' into functorch-rearrange

ringohoffman commented:
@zdevito lmk what you think of the latest changes!

zdevito commented on functorch/einops/_parsing.py:
```diff
@@ -0,0 +1,264 @@
+"""Adapted from https://github.com/arogozhnikov/einops/blob/36c7bb16e57d6e57f8f3050f9e07abdf3f00469f/einops/parsing.py.
+
```
Yep, this is fine

zdevito reviewed (approved):
Looks good! Thanks for making the changes.

zdevito commented:
@pytorchbot merge
